openrr-plugin/src/lib.rs

```diff
@@ -0,0 +1,519 @@
+
+mod proxy;
+
+use std::{
+    convert::TryInto,
+    fmt,
+    path::Path,
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+
+use abi_stable::{erased_types::TU_Opaque, library::lib_header_from_path, StableAbi};
+use anyhow::Result;
+use arci::{async_trait, Isometry2, Isometry3, WaitFuture};
+use auto_impl::auto_impl;
+
+// This is not a public API. Use export_plugin! macro for plugin exporting.
+#[doc(hidden)]
+pub use crate::proxy::PluginMod_Ref;
+use crate::proxy::{
+    GamepadTraitObject, JointTrajectoryClientTraitObject, LocalizationTraitObject,
+    MoveBaseTraitObject, NavigationTraitObject, PluginTraitObject, SpeakerTraitObject,
+    TransformResolverTraitObject,
+};
+
+/// Exports the plugin that will instantiated with the specified expression.
+///
+/// # Examples
+///
+/// ```
+/// use openrr_plugin::Plugin;
+///
+/// openrr_plugin::export_plugin!(MyPlugin);
+///
+/// pub struct MyPlugin;
+///
+/// impl Plugin for MyPlugin {
+///     fn name(&self) -> String {
+///         "MyPlugin".into()
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! export_plugin {
+    ($plugin_constructor:expr $(,)?) => {
+        /// Exports the root module of this library.
+        ///
+        /// This code isn't run until the layout of the type it returns is checked.
+        #[::abi_stable::export_root_module]
+        pub fn instantiate_root_module() -> $crate::PluginMod_Ref {
+            $crate::PluginMod_Ref::new(plugin_constructor)
+        }
+
+        /// Instantiates the plugin.
+        #[::abi_stable::sabi_extern_fn]
+        pub fn plugin_constructor() -> $crate::PluginProxy {
+            $crate::PluginProxy::new($plugin_constructor)
+        }
+    };
+}
+
+/// The plugin manager.
+#[derive(Debug, Default)]
+pub struct PluginManager {
+    plugins: Vec<PluginProxy>,
+}
+
+impl PluginManager {
+    /// Creates a new `PluginManager`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Loads a plugin from the specified path.
+    pub fn load(&mut self, path: impl AsRef<Path>) -> Result<()> {
+        let path = path.as_ref();
+
+        let header = lib_header_from_path(&path)?;
+        let root_module = header.init_root_module::<PluginMod_Ref>()?;
+
+        let plugin_constructor = root_module.plugin_constructor();
+        let plugin = plugin_constructor();
+
+        self.plugins.push(plugin);
+        Ok(())
+    }
+
+    /// Returns the list of all plugins managed by this plugin manager.
+    pub fn plugins(&self) -> &[PluginProxy] {
+        &self.plugins
+    }
+}
+
+/// The plugin trait.
+pub trait Plugin: 'static {
+    /// Returns the name of this plugin.
+    ///
+    /// NOTE: This is *not* a unique identifier.
+    fn name(&self) -> String;
+
+    /// Creates a new instance of [`arci::JointTrajectoryClient`] with the specified arguments.
+    fn new_joint_trajectory_client(
+        &self,
+        args: String,
+    ) -> Option<Box<dyn StaticJointTrajectoryClient>> {
+        drop(args);
```

Comment:
Just a question as a rust beginner.
Why we need drop?

Reply:
- Without it, `args` will be warned as an unused variable.
- There are two major ways to silence that warning, `#[allow(..)]` and leading `_`, but the former is not nice to look at and the latter is not nice to look when displayed in the documentation. Also, rust-analyzer generates method templates based on the trait definitions, so the user has to start writing by removing the leading `_`.

As a workaround for those problems, `drop` or `let _ =` sometimes be used in the default method for public traits.

- https://docs.rs/syn/1.0.72/src/syn/parse.rs.html#580-583
- https://github.com/serde-rs/serde/blob/1094e2d334f12705957f579f49adfc6118a1de1f/serde/src/ser/mod.rs#L1948
- https://github.com/tokio-rs/valuable/blob/c3efa86b4ac919fc14fe642eb5dfa55916e46e6f/valuable/src/visit.rs#L6

Reply:
Switched from `drop` to `let _ =` because that seems to be a more widely used pattern for that purpose.

Reply:
I see, thanks!